timescaledb_toolkit: init at 1.14.0

typetetris opened the pull request:
1. provide the function `buildPgxExtension` to build postgresql extensions with `cargo-pgx`
2. provide timescaledb_toolkit extension to postgresql build via `buildPgxExtension`
3. provide nixos tests making sure resulting timescaledb_toolkit extension can be loaded and used via postgresql nixos module

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [x] Tested, as applicable:
  - [nixos test](./nixos/tests/timescaledb.nix) at least for postgresql_14
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.11 Release Notes (or backporting 22.05 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2211-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

typetetris commented:
@winterqt @Hoverbear Maybe you want to take a look again.

SuperSandro2000 commented on pkgs/development/tools/rust/cargo-pgx/buildPgxExtension.nix:
```diff
@@ -0,0 +1,177 @@
+  "buildPgxExtensions overwrites the buildPhase, so providing one is useless";
+
+let
+  maybeDebugFlag = if buildType != "release" then "--debug" else "";
+  maybeEnterBuildAndTestSubdir =
+    if buildAndTestSubdir != null then ''
+      export CARGO_TARGET_DIR="$(pwd)/target"
+      pushd ${buildAndTestSubdir}
+    '' else "";
+  maybeLeaveBuildAndTestSubdir = if buildAndTestSubdir != null then "popd" else "";
```
```suggestion
  maybeDebugFlag = lib.optionalString (buildType != "release") "--debug";
  maybeEnterBuildAndTestSubdir = lib.optionalString (buildAndTestSubdir != null) ''
    export CARGO_TARGET_DIR="$(pwd)/target"
    pushd ${buildAndTestSubdir}
  '';
  maybeLeaveBuildAndTestSubdir = lib.optionalString (buildAndTestSubdir != null) "popd";
```

SuperSandro2000 commented on pkgs/development/tools/rust/cargo-pgx/buildPgxExtension.nix:
```diff
@@ -0,0 +1,177 @@
+    '' else "";
+  maybeLeaveBuildAndTestSubdir = if buildAndTestSubdir != null then "popd" else "";
+
+  pgxPostgresMajor = builtins.head (lib.splitString "." postgresql.version);
```
```suggestion
  pgxPostgresMajor = lib.versions.major postgresql.version;
```

SuperSandro2000 commented on pkgs/development/tools/rust/cargo-pgx/buildPgxExtension.nix:
```diff
@@ -0,0 +1,177 @@
+    mkdir -p $PGX_HOME/${pgxPostgresMajor}
+
+    cp -r -L ${postgresql}/. $PGX_HOME/${pgxPostgresMajor}/
+    chmod -R ugo+w $PGX_HOME/${pgxPostgresMajor}
```
```suggestion
    chmod -R +w $PGX_HOME/${pgxPostgresMajor}
```

SuperSandro2000 commented on pkgs/development/tools/rust/cargo-pgx/buildPgxExtension.nix:
```diff
@@ -0,0 +1,177 @@
+    chmod -R ugo+w $PGX_HOME/${pgxPostgresMajor}
+    cp -r -L ${postgresql.lib}/lib/. $PGX_HOME/${pgxPostgresMajor}/lib/
+
+    ${cargo-pgx}/bin/cargo-pgx pgx init \
```
```suggestion
    cargo-pgx pgx init \
```

cargo-pgx must be in nativeBuildInputs or cross compiling does not work.

SuperSandro2000 commented on pkgs/development/tools/rust/cargo-pgx/buildPgxExtension.nix:
```diff
@@ -0,0 +1,177 @@
+    export NIX_PGLIBDIR=$PGX_HOME/${pgxPostgresMajor}/lib
+
+    echo "unix_socket_directories = '$(mktemp -d)'" > $PGDATA/postgresql.conf
+    ${postgresql}/bin/pg_ctl start
+    ${postgresql}/bin/createuser -h localhost --superuser --createdb $USER || true
+    ${postgresql}/bin/pg_ctl stop
```
```suggestion
    pg_ctl start
    createuser -h localhost --superuser --createdb $USER || true
    pg_ctl stop
```
same for postgres

SuperSandro2000 commented on pkgs/development/tools/rust/cargo-pgx/buildPgxExtension.nix:
```diff
@@ -0,0 +1,177 @@
+    nativeBuildInputs = (args.nativeBuildInputs or [ ]) ++ [
+      cargo-pgx
+      rustc
+      rustfmt
```
```suggestion
```
Why do we need to format something? We shouldn't do this to prevent regressions when updating rustfmt.

typetetris replied:
`cargo-pgx` expects `rustfmt` to be in PATH, sadly

SuperSandro2000 commented on pkgs/development/tools/rust/cargo-pgx/buildPgxExtension.nix:
```diff
@@ -0,0 +1,177 @@
+      rustPlatform.bindgenHook
+    ];
+
+    postPatch = ''
+      ${args.postPatch or ""}
+      patchShebangs .
+    '';
```
```suggestion
    postPatch = args.postPatch or "" + ''
      patchShebangs .
    '';
```
Also what files need to be patched? Patching all files is rather slow.

typetetris replied:
I copied and pasted this from cargo-pgx example code. I'll investigate, wether its really needed like that. Otherwise, it could simply be erased I assume and considered a responsibility of the caller.

typetetris replied:
I think it is not necessary at all and if someone needs it, he can set `postPatch` himself. I will remove it.

SuperSandro2000 commented on pkgs/development/tools/rust/cargo-pgx/buildPgxExtension.nix:
```diff
@@ -0,0 +1,177 @@
+      export NIX_PGLIBDIR=${postgresql.out}/lib
+
+      PGX_BUILD_FLAGS="--frozen -j $NIX_BUILD_CORES ${builtins.concatStringsSep " " cargoBuildFlags}" \
+      ${cargo-pgx}/bin/cargo-pgx pgx package \
```
```suggestion
      cargo-pgx pgx package \
```

SuperSandro2000 commented on pkgs/development/tools/rust/cargo-pgx/buildPgxExtension.nix:
```diff
@@ -0,0 +1,177 @@
+      export NIX_PGLIBDIR=$PGX_HOME/${pgxPostgresMajor}/lib
+
+      ${maybeLeaveBuildAndTestSubdir}
+      echo "Finished cargo-pgx buildPhase"
```
```suggestion
```
That a phase is finished is usually indicated by the next one starting.

SuperSandro2000 commented on pkgs/development/tools/rust/cargo-pgx/buildPgxExtension.nix:
```diff
@@ -0,0 +1,177 @@
+      runHook postBuild
+    '';
+
+    preCheck = ''
+      ${preBuildAndTest}
+      ${args.preCheck or ""}
+    '';
```
```suggestion
    preCheck = preBuildAndTest
      + args.preCheck or "";
```

SuperSandro2000 commented on pkgs/development/tools/rust/cargo-pgx/buildPgxExtension.nix:
```diff
@@ -0,0 +1,177 @@
+
+      ${cargo-pgx}/bin/cargo-pgx pgx stop all
+
+      mv -v $out/${postgresql.out}/* $out
+      rm -rfv $out/nix
```
```suggestion
      mv $out/${postgresql.out}/* $out
      rm -rf $out/nix
```
We don't need to spam the log.

typetetris commented on pkgs/development/tools/rust/cargo-pgx/buildPgxExtension.nix:
```diff
@@ -0,0 +1,177 @@
+
+    checkNoDefaultFeatures = true;
+    checkFeatures = (args.checkFeatures or [ ]) ++ [ "pg_test pg${pgxPostgresMajor}" ];
+
+    shellHook = ''
+      echo 'ATTENTION'
+      echo 'to run the buildPhase'
+      echo '  - ensure port 5432 is free and can be used by postgres'
+      echo '  - run `eval "$buildPhase"`'
+      ${args.shellHook or ""}
+    '';
```
The user of `buildPgxExtension` should be notified about this somehow. I'll keep it as a comment at least.

typetetris commented on pkgs/servers/sql/postgresql/ext/timescaledb_toolkit.nix:
```diff
@@ -0,0 +1,41 @@
+, nixosTests
+}:
+cargo-pgx.buildPgxExtension rec {
+  inherit postgresql;
+
```
`buildPgxExtension` expects an argument `postgresql`, which is used a lot in `buildPgxExtension`.

SuperSandro2000 commented:
@marsam can you please take a look at this?

Also please squash the review commits to follow the contributing guide.

typetetris commented:
Cross compilation still fails:
```
at 19:30:46 ❯ nix-build . --arg crossSystem '{config="armv7l-unknown-linux-gnueabihf"; }' -A postgresqlPackages.timescaledb_toolkit
this derivation will be built:
  /nix/store/4g9k3gd471m8159z2r598h61zpmlampn-timescaledb_toolkit-armv7l-unknown-linux-gnueabihf-1.11.0.drv
building '/nix/store/4g9k3gd471m8159z2r598h61zpmlampn-timescaledb_toolkit-armv7l-unknown-linux-gnueabihf-1.11.0.drv'...
unpacking sources
unpacking source archive /nix/store/n2dy6l5in80y08124w638fwqzznbrrwg-source
source root is source
Executing cargoSetupPostUnpackHook
unpacking source archive /nix/store/yfqjw70f43d50q6s5cshahwksc3zl75n-timescaledb_toolkit-1.11.0-vendor.tar.gz
Finished cargoSetupPostUnpackHook
patching sources
Executing cargoSetupPostPatchHook
Validating consistency between /build/source/Cargo.lock and /build/timescaledb_toolkit-1.11.0-vendor.tar.gz/Cargo.lock
Finished cargoSetupPostPatchHook
updateAutotoolsGnuConfigScriptsPhase
configuring
building
Executing cargo-pgx buildPhase
/nix/store/l43k9kqpwdr2a85d278www19k4s6609g-stdenv-linux/setup: line 1403: /nix/store/cdkf0n2g9zf7vknbjp4j4di503dkq3y4-cargo-pgx-armv7l-unknown-linux-gnueabihf-0.4.5/bin/cargo-pgx: cannot execute binary file: Exec format error
error: builder for '/nix/store/4g9k3gd471m8159z2r598h61zpmlampn-timescaledb_toolkit-armv7l-unknown-linux-gnueabihf-1.11.0.drv' failed with exit code 126;
       last 10 log lines:
       > Finished cargoSetupPostUnpackHook
       > patching sources
       > Executing cargoSetupPostPatchHook
       > Validating consistency between /build/source/Cargo.lock and /build/timescaledb_toolkit-1.11.0-vendor.tar.gz/Cargo.lock
       > Finished cargoSetupPostPatchHook
       > updateAutotoolsGnuConfigScriptsPhase
       > configuring
       > building
       > Executing cargo-pgx buildPhase
       > /nix/store/l43k9kqpwdr2a85d278www19k4s6609g-stdenv-linux/setup: line 1403: /nix/store/cdkf0n2g9zf7vknbjp4j4di503dkq3y4-cargo-pgx-armv7l-unknown-linux-gnueabihf-0.4.5/bin/cargo-pgx: cannot execute binary file: Exec format error
       For full logs, run 'nix log /nix/store/4g9k3gd471m8159z2r598h61zpmlampn-timescaledb_toolkit-armv7l-unknown-linux-gnueabihf-1.11.0.drv'.
```
it somehow uses `cargo-pgx` compiled for the host platform, despite it being listed in `nativeBuildInputs` and should be used in the version compiled for the build platform, or not? I have to admit, I don't understand that very well.

typetetris commented:
Here are some evil hacks with which cargo-pgx/postgresql run at build time, but it now fails at a different place:

Code: https://github.com/typetetris/nixpkgs/commit/fa14c5e93921d4e8d39e8cc1d82f6efc9f817c22

Failure:

```
   Compiling pgx-utils v0.4.5                                                                                                                                 
   Compiling nalgebra v0.27.1                                                                                                                                 
   Compiling pgx-pg-sys v0.4.5                                                                                                                                
   Compiling pgx-macros v0.4.5                                                                                                                                
error: failed to run custom build command for `pgx-pg-sys v0.4.5`                                                                                             
                                                                                                                                                              
Caused by:                                                                                                                                                    
  process didn't exit successfully: `/build/source/target/release/build/pgx-pg-sys-525c36ceb182b0cd/build-script-build` (exit status: 101)                    
  --- stdout                                                                                                                                                  
  cargo:rerun-if-env-changed=PGX_PG_SYS_GENERATE_BINDINGS_FOR_RELEASE                                                                                         
  cargo:rerun-if-changed=/build/tmp.HYSAByhqtc/config.toml                                                                                                    
  cargo:rerun-if-changed=include/pg10.h                                                                                                                       
  cargo:rerun-if-changed=include/pg11.h                                        
  cargo:rerun-if-changed=include/pg12.h                                                                                                                       
  cargo:rerun-if-changed=include/pg13.h                                                                                                                       
  cargo:rerun-if-changed=include/pg14.h                                                                                                                       
  cargo:rerun-if-changed=cshim/pgx-cshim.c                                                                                                                    
  cargo:rerun-if-changed=cshim/Makefile                                                                                                                       
                                                                                                                                                              
  --- stderr                                                                                                                                                  
  manifest_dir=/build/timescaledb_toolkit-1.11.0-vendor.tar.gz/pgx-pg-sys                                                                                     
  shim_src=/build/timescaledb_toolkit-1.11.0-vendor.tar.gz/pgx-pg-sys/cshim                                                                                   
  shim_dst=/build/source/target/release/build/pgx-pg-sys-2320b93506f5120f/out/cshim                                                                           
  Generating bindings for pg14                                                                                                                                
  /nix/store/p6knl0hcfah9myg5xzn5kxd1kjp3fiy5-glibc-armv7l-unknown-linux-gnueabihf-2.35-163-dev/include/gnu/stubs.h:7:11: fatal error: 'gnu/stubs-soft.h' file not found                                                                                                                                                    
  /nix/store/p6knl0hcfah9myg5xzn5kxd1kjp3fiy5-glibc-armv7l-unknown-linux-gnueabihf-2.35-163-dev/include/gnu/stubs.h:7:11: fatal error: 'gnu/stubs-soft.h' file not found, err: true                                                                                                                                         
  thread 'main' panicked at 'Unable to generate bindings for pg14: ()', /build/timescaledb_toolkit-1.11.0-vendor.tar.gz/pgx-pg-sys/build.rs:530:13            
  stack backtrace:                                                                                                                                            
     0:     0x555555e3898c - <std::sys_common::backtrace::_print::DisplayBacktrace as core::fmt::Display>::fmt::hf2ea00d5f3b53d9f                             
     1:     0x555555e6a3de - core::fmt::write::ha076a7e65c6b1874                                                                                              
     2:     0x555555e4a671 - std::io::Write::write_fmt::h9630bb3b0fe886f5                                                                                                                                                                                                                                                   
     3:     0x555555e52a45 - std::panicking::default_hook::{{closure}}::h64163197f96398c1                                                                     
     4:     0x555555e526a2 - std::panicking::default_hook::h92625161492e6b48       
```

Again I am at a loss.

typetetris committed:
cargo-pgx: add buildPgxExtension

typetetris committed:
timescaledb-toolkit: init at 1.14.0

typetetris commented:
> Also please squash the review commits to follow the contributing guide.

Partially done. Kept two different commits for cargo-pgx changes and adding timescaledb_toolkit. I would have done two pull requests, but the second one would bring the first tests for the first one, which seems awkward.

SuperSandro2000 commented:
Doing this in one PR is fine. The nixos test should also get its own commit.

I think the changes look good from someone that has never worked with postgres plugins.

typetetris committed:
cargo-pgx/timescaledb_toolkit: add nixos test

typetetris commented:
> The nixos test should also get its own commit.

Done.

I also fixed the comment on `buildPgxExtension` which claimed `postgresql` was the only argument in addition to all the arguments for `buildRustPackage`.

typetetris commented:
@SuperSandro2000
@marsam 

Updated to 1.13.1 so cargo-pgx 0.5.4 isn't needed any more.

This is open now for quite some time.
What do I need to change, so it can be merge?
Can it be merged at all?

marsam commented:
Thank a lot for working on this and sorry for the delay. 

Unfortunately, I've been awful busy with life and work. I'm hoping to review this PR the next week.

SuperSandro2000 commented on pkgs/servers/sql/postgresql/ext/timescaledb_toolkit.nix:
```diff
@@ -0,0 +1,42 @@
+    license = licenses.asl20;
+
+    # as it needs to be used with timescaledb, simply use the condition from there
+    broken = versionOlder postgresql.version "12";
```
If that is essentially copied, would it make sense to use timescaledb.meta.broken?

typetetris replied:
`timescaledb` and `timescaledb-toolkit` have different version schemes. So it might happen, the latest version of `timescaledb` does support a new postgresql version, but the latest version of `timescaldb-toolkit` doesn't. And the other way round. But yes, it will most likely be short periods of time.

typetetris replied:
I would need to change (in `pkgs/servers/sql/postgresql/packages.nix`)
`timescaledb_toolkit = super.callPackage ./ext/timescaledb_toolkit.nix { };`
to
`timescaledb_toolkit = self.callPackage ./ext/timescaledb_toolkit.nix { };`
to get access to `timescaledb` and it would be the first extension to do so.

So I would like to avoid that. A tiny bit of duplication is better than an additional
dependency in my mind.

typetetris commented:
Rebased on master.

marsam commented on pkgs/development/tools/rust/cargo-pgx/buildPgxExtension.nix:
```diff
@@ -0,0 +1,161 @@
+# Additional arguments are:
+#   - `postgresql` postgresql package of the version of postgresql this extension should be build for.
+#                  Needs to be the build platform variant.
+#   - `useFakeRustfmt` Wether to use a noop fake command as rustfmt. cargo-pgx tries to call rustfmt.
```
```suggestion
#   - `useFakeRustfmt` Whether to use a noop fake command as rustfmt. cargo-pgx tries to call rustfmt.
```

marsam commented on pkgs/development/tools/rust/cargo-pgx/buildPgxExtension.nix:
```diff
@@ -0,0 +1,161 @@
+    export PGX_HOME=$(mktemp -d)
+    export PGDATA="$PGX_HOME/data-${pgxPostgresMajor}/"
+    cargo-pgx pgx init "--pg${pgxPostgresMajor}" ${postgresql}/bin/pg_config
+    echo "unix_socket_directories = '$(mktemp -d)'" > "$PGDATA/postgresql.conf"
```
I think you also need to disable the tcp ports (`listen_addresses = ''`) to fix the darwin build, because on darwin they usually don't have sandbox enabled by default

typetetris replied:
Ok.

marsam commented on pkgs/development/tools/rust/cargo-pgx/buildPgxExtension.nix:
```diff
@@ -0,0 +1,161 @@
+    export USER="$(whoami)"
+    pg_ctl start
+    createuser -h localhost --superuser --createdb "$USER" || true
+    pg_ctl stop
```
Can we move the pg_ctl stop to the `postCheck` phase? I'm thinking of starting the server in preBuild and shutting it down in postCheck.

typetetris replied:
The server is only started for the `createuser` command. Afterwards it is not started again by the build script. I would have to check, whether it is started by `cargo pgx`.

marsam commented on pkgs/development/tools/rust/cargo-pgx/buildPgxExtension.nix:
```diff
@@ -0,0 +1,161 @@
+    nativeBuildInputs = (args.nativeBuildInputs or [ ]) ++ [
+      cargo-pgx
+      postgresql
+      rustPlatform.rust.rustc
```
rustPlatform already adds [rustc to nativeBuildInputs](https://github.com/NixOS/nixpkgs/blob/82b4c24705340bceed0d463d3d897b6ebcbd7550/pkgs/build-support/rust/build-rust-package/default.nix#L132)
```suggestion
```

marsam commented on pkgs/development/tools/rust/cargo-pgx/buildPgxExtension.nix:
```diff
@@ -0,0 +1,161 @@
+      cargo-pgx
+      postgresql
+      rustPlatform.rust.rustc
+      pkg-config
```
just curious, is pkg-config needed?

typetetris replied:
IIRC `cargo pgx` calls pkgs-config, but I will check this and remove it, if it isn't needed.

marsam commented on nixos/tests/all-tests.nix:
```diff
@@ -683,6 +683,7 @@ in {
   terminal-emulators = handleTest ./terminal-emulators.nix {};
   tiddlywiki = handleTest ./tiddlywiki.nix {};
   tigervnc = handleTest ./tigervnc.nix {};
+  timescaledb = handleTest ./timescaledb.nix {};
```
what do you think of renaming this to `timescaledb-toolkit`?
```suggestion
  timescaledb-toolkit = handleTest ./timescaledb-toolkit.nix {};
```

typetetris replied:
There was no test for timescaledb already, so this test is a very minimal test for both things, timescaledb and timescaledb-toolkit.
But I am not attached to the name. I will go ahead and change this.

marsam commented on pkgs/development/tools/rust/cargo-pgx/buildPgxExtension.nix:
```diff
@@ -0,0 +1,161 @@
+      NIX_PGLIBDIR="${postgresql}/lib" \
+      PGX_BUILD_FLAGS="--frozen -j $NIX_BUILD_CORES ${builtins.concatStringsSep " " cargoBuildFlags}" \
+      cargo-pgx pgx package \
+      --pg-config ${postgresql}/bin/pg_config \
+      ${maybeDebugFlag} \
+      --features "${builtins.concatStringsSep " " buildFeatures}" \
+      --out-dir "$out"
```
nitpick 
```suggestion
        --pg-config ${postgresql}/bin/pg_config \
        ${maybeDebugFlag} \
        --features "${builtins.concatStringsSep " " buildFeatures}" \
        --out-dir "$out"
```

marsam commented on pkgs/development/tools/rust/cargo-pgx/buildPgxExtension.nix:
```diff
@@ -0,0 +1,161 @@
+
+    nativeBuildInputs = (args.nativeBuildInputs or [ ]) ++ [
+      cargo-pgx
+      postgresql
```
I think postgresql should be in `buildInputs`

typetetris replied:
Hmm. Maybe timescaledb-toolkit isn't easily cross compiled. `cargo pgx` wants to call `pg_config` so we need the postgresql derivation having the timescaledb-toolkits build platform as host platform.

But on the other hand, it builds a library, which hopefully isn't linked against anything from that postgresql package, but just uses the header files from that package and declares some symbols to be resolved at runtime, so it should work against postgresql compiled for a different platform.

So timescaledb-toolkit shouldn't use anything from that postgresql derivation at runtime.

That is just my understanding.

I will test that, if I can. (Don't you need to by a mac to do darwin builds?)

marsam reviewed (commented):
Sorry for the delay. LGTM overall, thanks a lot for working on this!

typetetris committed:
Apply suggestions from code review

Typos, Whitespace and Stuff. Will fixup the relevants commits later, so this commit vanishes.

Co-authored-by: Mario Rodas <marsam@users.noreply.github.com>

marsam reviewed (approved):
LGTM. Thank a lot for working on this! 
If there is no other reviews, I'll merge it this weekend.
